src/main/java/net/snowflake/ingest/connection/OAuthManager.java

```diff
@@ -0,0 +1,241 @@
+
+import static net.snowflake.ingest.utils.Constants.*;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import net.snowflake.client.jdbc.internal.apache.http.HttpHeaders;
+import net.snowflake.client.jdbc.internal.apache.http.client.methods.CloseableHttpResponse;
+import net.snowflake.client.jdbc.internal.apache.http.client.methods.HttpPost;
+import net.snowflake.client.jdbc.internal.apache.http.entity.ContentType;
+import net.snowflake.client.jdbc.internal.apache.http.entity.StringEntity;
+import net.snowflake.client.jdbc.internal.apache.http.impl.client.CloseableHttpClient;
+import net.snowflake.client.jdbc.internal.apache.http.util.EntityUtils;
+import net.snowflake.client.jdbc.internal.google.api.client.http.HttpStatusCodes;
+import net.snowflake.client.jdbc.internal.google.gson.JsonObject;
+import net.snowflake.client.jdbc.internal.google.gson.JsonParser;
+import net.snowflake.ingest.utils.*;
+
+/**
+ * This class manages creating and automatically renewing the JWT token
+ *
+ * @author obabarinsa
+ * @since 1.8
+ */
+final class OAuthManager extends SecurityManager {
+
+  // Default update threshold, a floating-point value representing the ratio between the expiration
+  // time of an access
+  // token and the time needed to update it. It must be a value greater than 0 and less than or
+  // equal to 1.
+  private static final double DEFAULT_UPDATE_THRESHOLD_RATIO = 0.8;
+  private static final String TOKEN_TYPE = "OAUTH";
+  private static final String OAUTH_CONTENT_TYPE_HEADER = "application/x-www-form-urlencoded";
+
+  private static final String GRANT_TYPE_PARAM = "grant_type";
+
+  private static final String ACCESS_TOKEN = "access_token";
+  private static final String REFRESH_TOKEN = "refresh_token";
+  private static final String EXPIRES_IN = "expires_in";
+  private final double updateThresholdRatio;
+
+  private final AtomicReference<OAuthCredential> oAuthCredential;
+
+  // Did we fail to regenerate our token at some point?
+  private final AtomicBoolean refreshFailed;
+
+  private final URI tokenRequestURI;
+
+  private final CloseableHttpClient httpClient;
+
+  // Thread factory for daemon threads so that application can shutdown
+  final ThreadFactory tf = ThreadFactoryUtil.poolThreadFactory(getClass().getSimpleName(), true);
+
+  // the thread we use for refreshing access token
+  private final ScheduledExecutorService tokenRefresher = Executors.newScheduledThreadPool(1, tf);
+
+  // Reference to the Telemetry Service in the client
+  private final TelemetryService telemetryService;
```

Comment:
Looks like these are in the JWT token manager as well - can we move them to a shared base class so we don't need to duplicate these?